app_dart/lib/src/request_handlers/check_for_waiting_pull_requests.dart

```diff
@@ -0,0 +1,309 @@
+  })  : loggingProvider = loggingProvider ?? Providers.serviceScopeLogger,
+        super(config: config, authenticationProvider: authenticationProvider);
+
+  final LoggingProvider loggingProvider;
+
+  @override
+  Future<Body> get() async {
+    final Logging log = loggingProvider();
+
+    if (authContext.clientContext.isDevelopmentEnvironment) {
+      // Don't push GitHub status from the local dev server.
+      return Body.empty;
+    }
+
+    await _checkPRs(log, RepositorySlug('flutter', 'flutter'));
+    await _checkPRs(log, RepositorySlug('flutter', 'engine'));
+
+    return Body.empty;
+  }
+
+  Future<void> _checkPRs(Logging log, RepositorySlug slug) async {
+    final Map<String, dynamic> data = await _queryGraphQL(slug, log);
+    for (_AutoMergeQueryResult queryResult in _parseQueryData(data)) {
+      if (queryResult.shouldMerge) {
+        await _mergePullRequest(
+          log,
+          RepositorySlug('flutter', 'flutter'),
+          queryResult.number,
+          queryResult.sha,
+        );
+      } else if (queryResult.shouldRemoveLabel) {
+        await _removeLabel(
+          slug,
+          queryResult.number,
+          queryResult.removalMessage,
+        );
+      } else {
+        log.debug('Pull Request could be merged but tree is red: $queryResult');
+      }
+    }
+  }
+
+  Future<Map<String, dynamic>> _queryGraphQL(RepositorySlug slug, Logging log) async {
+    final HttpLink _httpLink = HttpLink(
+      uri: 'https://api.github.com/graphql',
+    );
+
+    final String token = await config.githubOAuthToken;
+    final AuthLink _authLink = AuthLink(
+      getToken: () async => 'Bearer $token',
+    );
+
+    final Link _link = _authLink.concat(_httpLink);
+
+    final GraphQLClient _client = GraphQLClient(
+      cache: InMemoryCache(),
+      link: _link,
+    );
+
+    final String graphQLDocument = await _graphQLDocumentForSlug(slug);
+
+    final QueryResult result = await _client.query(
+      QueryOptions(
+        document: graphQLDocument,
+        fetchPolicy: FetchPolicy.noCache,
+      ),
+    );
+
+    if (result.hasErrors) {
+      log.error(jsonEncode(result.errors));
+      throw const InternalServerError();
+    }
+
+    return result.data;
+  }
+
+  Future<String> _graphQLDocumentForSlug(RepositorySlug slug) async {
+    final String labelName = await config.waitingForTreeToGoGreenLabelName;
+    return '''query {
+repository(owner: "${slug.owner}", name: "${slug.name}") {
+  pullRequests(labels: "$labelName", first: 100, states: OPEN) {
+    nodes {
+      number
+      mergeable
+      commits(last:1) {
+        nodes {
+          commit {
+            abbreviatedOid
+            oid
+            status {
+              state
+            }
+          }
+        }
+      }
+      reviews(first: 100, states: [CHANGES_REQUESTED, APPROVED]) {
+        nodes {
+          state
+          author {
+            login
+          }
+        }
+      }
+    }
+  }
+}}''';
```

Comment:
what dark magic is this? :-o

Reply:
GraphQL query.  This lets us get reviews for specifically labeled PRs in a single request, instead of making a request for PR, separate request for its reviews, and then joining/filtering all of those client side.